Bottom: be1adbff266685f93dcf753fb5d3ee0a98f7aa34
Top:    65d2172e4a1eb2f12744515f8cc61509bbb7fd0f
Author: Alexey Lyashkov <shadow@Alexeys-MacBook-Pro.local>
Date:   2016-01-16 21:19:26 +0300

Refresh of ctl-mgs.patch

---

diff --git a/Makefile b/Makefile
index 590400a..313194f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,7 @@
 ifneq ($(KERNELRELEASE),)
     obj-m	:= netsrv.o netcli.o
     obj_ib-y	:= ib-sock.o ib-sock-util.o
+    obj_ib-y	+= ib-sock-mem.o
     netsrv-objs	:= srv.o $(obj_ib-y)
     netcli-objs	:= cli.o $(obj_ib-y)
 else
diff --git a/ib-sock-int.h b/ib-sock-int.h
index a934272..14179f7 100644
--- a/ib-sock-int.h
+++ b/ib-sock-int.h
@@ -19,6 +19,19 @@ enum ib_sock_flags {
 	SOCK_ERROR	= 1 << 1,
 };
 
+struct ib_sock_mem {
+	/* protection domain */
+	struct ib_pd		*ism_pd;
+	/* memory window to map.
+	 * all ? or most cards may work with single == global MR  */
+	struct ib_mr		*ism_mr;
+	
+	/* # send work items for single transfer */
+	unsigned		ism_wr_count;
+	/* ...and their memory */
+	unsigned		ism_sge_count;
+};
+
 struct IB_SOCK {
 	/* primary OFED stack ID */
 	struct rdma_cm_id	*is_id;
diff --git a/ib-sock-mem.c b/ib-sock-mem.c
new file mode 100644
index 0000000..54d8230
--- /dev/null
+++ b/ib-sock-mem.c
@@ -0,0 +1,56 @@
+#include "ib-sock-int.h"
+
+/* it may done for device only (not per connection base) */
+static int ib_sock_mem_init_common(struct IB_SOCK *sock)
+{
+	int 	ret;
+	int	access_flags = IB_ACCESS_LOCAL_WRITE |
+				IB_ACCESS_REMOTE_WRITE;
+
+	/* create a protection domain. */
+	sock->is_mem.ism_pd = ib_alloc_pd(sock->is_id->device);
+	if (IS_ERR(sock->is_mem.ism_pd)) {
+		printk("can't create a protection domain\n");
+		ret = PTR_ERR(sock->is_mem.ism_pd);
+		sock->is_mem.ism_pd = NULL;
+		return ret;
+	}
+
+	sock->is_mem.ism_mr = ib_get_dma_mr(sock->is_mem.ism_pd, access_flags);
+	if (IS_ERR(sock->is_mem.ism_mr)) {
+		ret = PTR_ERR(sock->is_mem.ism_mr);
+		sock->is_mem.ism_mr = NULL;
+		printk("Failed ib_get_dma_mr : %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void ib_sock_mem_fini_common(struct IB_SOCK *sock)
+{
+	if (sock->is_mem.ism_mr)
+		ib_dereg_mr(sock->is_mem.ism_mr);
+
+	if (sock->is_mem.ism_pd)
+		ib_dealloc_pd(sock->is_mem.ism_pd);
+}
+
+int ib_sock_mem_init(struct IB_SOCK *sock)
+{
+	int i;
+	int ret;
+
+	ret = ib_sock_mem_init_common(sock);
+	if (ret < 0)
+		return ret;
+
+	/* different memory managment models may need different init */
+
+	return 0;
+}
+
+void ib_sock_mem_fini(struct IB_SOCK *sock)
+{
+	ib_sock_mem_fini_common(sock);
+}
